File: pt4/1054.distant-barcodes.go
package main

import "container/heap"

/*
1054. Distant Barcodes
*/
func rearrangeBarcodes(barcodes []int) []int {
	counts := make(map[int]int)
	for _, barcode := range barcodes {
		counts[barcode]++
	}

	items := make(barcodeMaxHeap, 0, len(counts))
	for barcode, count := range counts {
		items = append(items, barcodeEntry{count: count, value: barcode})
	}
	heap.Init(&items)

	result := make([]int, 0, len(barcodes))
	var previous barcodeEntry
	hasPrevious := false

	for items.Len() > 0 {
		current := heap.Pop(&items).(barcodeEntry)
		result = append(result, current.value)
		current.count--

		if hasPrevious && previous.count > 0 {
			heap.Push(&items, previous)
		}

		previous = current
		hasPrevious = true
	}

	return result
}

type barcodeEntry struct {
	count int
	value int
}

type barcodeMaxHeap []barcodeEntry

func (h barcodeMaxHeap) Len() int { return len(h) }

func (h barcodeMaxHeap) Less(i, j int) bool {
	if h[i].count == h[j].count {
		return h[i].value < h[j].value
	}
	return h[i].count > h[j].count
}

func (h barcodeMaxHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }

func (h *barcodeMaxHeap) Push(x interface{}) {
	*h = append(*h, x.(barcodeEntry))
}

func (h *barcodeMaxHeap) Pop() interface{} {
	old := *h
	item := old[len(old)-1]
	*h = old[:len(old)-1]
	return item
}

/*
Interview Explanation

Core idea:
Always place the most frequent barcode that is not equal to the one just
placed. Holding the previous barcode out of the heap for one step enforces the
adjacency rule.

Go data structures:
- map[int]int counts barcode frequencies.
- container/heap is used with a custom max-heap type. Go's heap package is a
  min-heap by default, so Less is reversed by count.
- previous stores the last used barcode until a different barcode is placed.

Algorithm:
1. Count frequencies.
2. Build a max heap of (count, barcode).
3. Pop the most frequent currently allowed barcode and append it.
4. Decrement its count.
5. Push the previous barcode back after placing a different one.

Correctness:
The last placed barcode is absent from the heap during the next selection, so
two adjacent positions are never equal. Since a valid answer is guaranteed,
choosing the most frequent available barcode keeps high-frequency values spread
out and completes a valid rearrangement.

Complexity:
Let n be len(barcodes) and k be the number of distinct barcodes. Time is
O(n log k), and extra space is O(k) besides the result.

Edge cases:
- Length one returns that single barcode.
- Equal frequencies alternate naturally.
- A dominant barcode is repeatedly delayed by one position.
*/

File: pt4/1054.distant-barcodes.py
#
# @lc app=leetcode id=1054 lang=python3
#
# [1054] Distant Barcodes
#
# https://leetcode.com/problems/distant-barcodes/description/
#
# algorithms
# Medium (48.90%)
# Likes:    1355
# Dislikes: 52
# Total Accepted:    58K
# Total Submissions: 118.5K
# Testcase Example:  '[1,1,1,2,2,2]'
#
# In a warehouse, there is a row of barcodes, where the i^th barcode is
# barcodes[i].
# 
# Rearrange the barcodes so that no two adjacent barcodes are equal. You may
# return any answer, and it is guaranteed an answer exists.
# 
# 
# Example 1:
# Input: barcodes = [1,1,1,2,2,2]
# Output: [2,1,2,1,2,1]
# Example 2:
# Input: barcodes = [1,1,1,1,2,2,3,3]
# Output: [1,3,1,3,1,2,1,2]
# 
# 
# Constraints:
# 
# 
# 1 <= barcodes.length <= 10000
# 1 <= barcodes[i] <= 10000
# 
# 
#

# @lc code=start
from collections import Counter
from heapq import heapify, heappop, heappush
from typing import List


class Solution:
    def rearrangeBarcodes(self, barcodes: List[int]) -> List[int]:
        heap = [(-count, barcode) for barcode, count in Counter(barcodes).items()]
        heapify(heap)

        result = []
        previous_count = 0
        previous_barcode = 0

        while heap:
            count, barcode = heappop(heap)
            result.append(barcode)
            count += 1

            if previous_count < 0:
                heappush(heap, (previous_count, previous_barcode))

            previous_count = count
            previous_barcode = barcode

        return result
# @lc code=end

"""
Interview Explanation

Core idea:
Always place the barcode with the largest remaining count that is not equal to
the barcode placed immediately before it. Holding the previous barcode out of
the heap for one step enforces the adjacency rule.

Algorithm:
1. Count barcode frequencies.
2. Store them in a max heap using negative counts.
3. Repeatedly pop the most frequent currently allowed barcode and append it.
4. Decrease its count.
5. Push the previous barcode back only after a different barcode has been
   placed.

Data structure choice:
A max heap gives quick access to the most frequent remaining barcode. The
single "previous" slot temporarily blocks the last placed value from being
chosen again immediately.

Correctness:
The algorithm never places the same barcode twice in a row because the last
placed barcode is not in the heap during the next selection. Choosing the most
frequent available barcode is safe because a valid answer is guaranteed, and
using high-frequency values early prevents them from being stranded at the end.

Complexity:
Let n be the number of barcodes and k the number of distinct values. Time is
O(n log k), and space is O(k) besides the result.

Tests and edge cases:
- One barcode: returns it.
- Two equally frequent barcodes alternate.
- One dominant barcode, such as [1,1,1,1,2,2,3,3], is spread out.
- The returned order may differ from examples; any valid arrangement is fine.
"""
